streaming/base/batching/stratified.py

```diff
@@ -0,0 +1,171 @@
+# Copyright 2023 MosaicML Streaming authors
+# SPDX-License-Identifier: Apache-2.0
+
+"""Apportion shards/samples so each batch has the same amount of samples from each stream."""
+from __future__ import annotations
+
+import logging
+from typing import TYPE_CHECKING
+
+import numpy as np
+from numpy.typing import NDArray
+
+from streaming.base.partition import get_partitions
+from streaming.base.shuffle import get_shuffle
+from streaming.base.world import World
+
+if TYPE_CHECKING:
+    from streaming.base.dataset import StreamingDataset
+
+logger = logging.getLogger(__name__)
+
+
+def generate_work_stratified_batching(dataset: StreamingDataset, world: World, epoch: int,
+                                      sample_in_epoch: int) -> NDArray[np.int64]:
+    """Generate the epoch's sample arrangement for ``stratified`` batching method.
+
+    This is only called in local rank zero. When ``batching_method`` is set to ``stratified``,
+    every single batch is divided between streams in the same proportions.
+
+    Args:
+        dataset (StreamingDataset): Dataset to generate the partition for.
+        world (World): World state.
+        epoch (int): Which epoch it is.
+        sample_in_epoch (int): Where we are in the epoch.
+
+    Returns:
+        NDArray[np.int64]: The epoch (num physical nodes, ranks per node, workers per rank,
+            batches per worker, batch size).
+    """
+    # Ensure that num_canonical_nodes has been set.
+    if dataset.num_canonical_nodes is None:
+        raise RuntimeError(f'`num_canonical_nodes` can never be None. ' +
+                           f'Provide a positive integer.')
+
+    # First, for each stream, sample each shard of the stream according to proportions/repeats/samples.
+    # We obtain the resampled size of each shard in the stream and a mapping from the training "big" sample ID
```

Comment:
is this under 100 character limit ? Please also check other places if it is applicable.

Reply:
fixed!

Reply:
i also fixed wherever this happened everywhere else in the codebase.